Add ProGit v1 redirects

dscho opened the pull request:
# Changes

With this PR, the original ProGit v1 links (like https://git-scm.com/book/en/Getting-Started-Git-Basics) will no longer show the 404 page, but instead redirect to the corresponding v2 section.

## Context

It was reported in https://github.com/git/git-scm.com/issues/1782 that the ProGit v1 URLs stopped working a while ago. This was pre-Hugo/Pagefind switch, resulting in a 500 page, but after that switch, they are still broken, only now a 400 page is shown instead.

I mulled for a while what to do because we cannot have wildcard routes in a static site.

Eventually, I suggested to resurrect the previously-working URLs by inspecting the original `progit` repository.

In the end, I figured out a way that was much more convenient: obtain the table of contents from the Internet Archive and then use common sense to manually map the v1 URLs to the v2 URLs.

I did not want to leave potential contributors with a lot of uncertainty whether this is even workable, so I investigated a bit, and even came up with that manual [v1 <-> v2 mapping.](https://github.com/git/git-scm.com/issues/1782#issuecomment-2457407947).

After that, I did not want to invest even more effort myself, hoping that other people would run with the idea. But nobody bit. So today, I thought: "How hard could this be?". And to my surprise, it was not hard at all.

Note that only the first 5 commits of this PR are interesting with regards to review; The remainder of the commits were created by [a manually-triggered `update-book` workflow run](https://github.com/dscho/git-scm.com/actions/runs/11705779665) where I specifically [forced](https://github.com/dscho/git-scm.com/actions/runs/11705779665/job/32601299950#step:3:5) a complete rebuild (the second parameter of the `getPendingBookUpdates()` call [reflects the `force-rebuild` input parameter](https://github.com/dscho/git-scm.com/actions/runs/11705779665/workflow#L30).

For convenience, I deployed this to my fork, you can verify that the ProGit v1 URLs work e.g. by picking a link from [the original ProGit v1 table of contents in the Internet Archive](https://web.archive.org/web/20140122054205/http://git-scm.com/book), editing it so that it points to my forked site instead, and then verify that it redirects as intended. For example:

- https://web.archive.org/web/20140122054205/http://git-scm.com/book/en/Git-on-the-Server-Generating-Your-SSH-Public-Key
- ➡️ https://dscho.github.io/git-scm.com/book/en/Git-on-the-Server-Generating-Your-SSH-Public-Key
- ➡️ https://dscho.github.io/git-scm.com/book/en/v2/Git-on-the-Server-Generating-Your-SSH-Public-Key.

I also tried to make sure that the translated pages work as well. Example:

- https://web.archive.org/web/20140122062500/http://git-scm.com/book/mk/%D0%9E%D1%81%D0%BD%D0%BE%D0%B2%D0%B8-%D0%BD%D0%B0-Git-%D0%97%D0%B0%D1%87%D1%83%D0%B2%D1%83%D0%B2%D0%B0%D1%9A%D0%B5-%D0%BD%D0%B0-%D0%BF%D1%80%D0%BE%D0%BC%D0%B5%D0%BD%D0%B8-%D0%B2%D0%BE-%D0%B2%D0%B0%D1%88%D0%B5%D1%82%D0%BE-%D1%80%D0%B5%D0%BF%D0%BE%D0%B7%D0%B8%D1%82%D0%BE%D1%80%D0%B8
- ➡️https://dscho.github.io/git-scm.com/book/mk/%D0%9E%D1%81%D0%BD%D0%BE%D0%B2%D0%B8-%D0%BD%D0%B0-Git-%D0%97%D0%B0%D1%87%D1%83%D0%B2%D1%83%D0%B2%D0%B0%D1%9A%D0%B5-%D0%BD%D0%B0-%D0%BF%D1%80%D0%BE%D0%BC%D0%B5%D0%BD%D0%B8-%D0%B2%D0%BE-%D0%B2%D0%B0%D1%88%D0%B5%D1%82%D0%BE-%D1%80%D0%B5%D0%BF%D0%BE%D0%B7%D0%B8%D1%82%D0%BE%D1%80%D0%B8
- ➡️https://dscho.github.io/git-scm.com/book/mk/v2/%D0%9E%D1%81%D0%BD%D0%BE%D0%B2%D0%B8%D1%82%D0%B5-%D0%BD%D0%B0-Git-%D0%A1%D0%BD%D0%B8%D0%BC%D0%B0%D1%9A%D0%B5-%D0%BD%D0%B0-%D0%BF%D1%80%D0%BE%D0%BC%D0%B5%D0%BD%D0%B8-%D0%B2%D0%BE-%D1%81%D0%BA%D0%BB%D0%B0%D0%B4%D0%B8%D1%88%D1%82%D0%B5%D1%82%D0%BE

This fixes https://github.com/git/git-scm.com/issues/1782

dscho commented:
I just updated this by rebasing to the current `gh-pages` revision (skipping the book updates), then letting [another `update-book` workflow run](https://github.com/dscho/git-scm.com/actions/runs/11729990852) update it. [Here](https://github.com/git/git-scm.com/compare/df06492edbf27cfe8b0d69dbe74ad3293c3dac4a..4b0016970bcd5ed5d01024d76d7343dbbe50cc8c) is the diff (spoiler alert: it essentially just updates the eBook download URLs).

LemmingAvalanche commented:
I was [asked](https://github.com/git/git-scm.com/issues/1782#issuecomment-2464636795) to review this PR, a task I’m not qualified for (no Ruby knowledge etc.). So I’ll probably try to read the code back.

LemmingAvalanche commented on script/extract-book-v1-urls.rb:
```diff
@@ -0,0 +1,79 @@
+require 'uri'
+require 'yaml'
+
+$v1_to_v2 = [
```
This must correspond to this table:

https://github.com/git/git-scm.com/issues/1782#issuecomment-2457407947

But how does this work? Are the values here the V2 target?  Is the V1 chapter mapped via a 2D index?  Like for V1 Ch. 1.4 Installing Git: V2 Ch. 1.5 is in the fourth position here so that seems to line up.

dscho replied:
This map has the v2 sections as keys, and maps to values that are arrays of v1 sections to which a particular v2 section needs to be aliased.

dscho replied:
> This map has the v2 sections as keys, and maps to values that are arrays of v1 sections to which a particular v2 section needs to be aliased.

Oh, scratch that, I was talking about `data/book_v1.yml`. Will add a code comment to clarify the role of `v1_to_v2`.

dscho replied:
Uh oh. The `$v1_to_v2` data is not even used. Which means that https://dscho.github.io/git-scm.com/book/en/Git-Tools-Submodules redirects to https://dscho.github.io/git-scm.com/book/en/v2/GitHub-Summary (i.e. 6.6 -> 6.6, when it should redirect to https://dscho.github.io/git-scm.com/book/en/v2/Git-Tools-Submodules instead, i.e. to 7.11)!

Will fix immediately.

LemmingAvalanche commented on script/extract-book-v1-urls.rb:
I like the approach of scraping the Internet for what’s in use out in the wild.

I got a few errors eventually when running this script (`Errno::ECONNREFUSED`).  But rerunning it enough times seemed to download everything. Apparently it remembers what it has already downloaded.

dscho replied:
Yes, it writes `cached.book-toc.<language>.html` files; I had the same `ECONNREFUSED` errors, that's why I introduced that.

dscho replied:
Given that this script is supposed to be used exactly once, to initialize `data/book_v1.yml` (which is then not supposed to change ever again, at least not until v3 if that ever materializes), I'd rather focus on more pressing issues ;-)

LemmingAvalanche commented on script/book.rb:
```diff
@@ -199,6 +203,19 @@ def save
     end
   end
+
+  def book_v1_aliases(cs_number)
```
So this follows the same style as the rest of this script.  And my reading of the script: this script generates all the static pages for the book. And deals with URLs.

In particular for redirects: we can see before this change that there is logic surrounding `front_matter` about aliases, and relevant to this PR, aliases to redirect from V1 links to the current book (V2).

Now this PR adds this function which is effectively a thin wrapper around the YAML files of the out-in-the-wild (Wayback Machine) URLs.

Makes sense.

LemmingAvalanche commented:
I wanted to test some redirects but I had the same local-only problem as last time I tried to serve the page locally: got infinite redirects when navigating.

LemmingAvalanche commented:
I compared the `book_v1.yml` on Chapter 4 (since V1/V2 diverge a lot) to Dscho’s initial table:

https://github.com/git/git-scm.com/issues/1782#issuecomment-2457407947

The mapping in that file is the same as the table.

LemmingAvalanche commented:
I was a bit confused by all the one-commit merge commits for the localized book versions.  But from reading the Git mailing list I know that Dscho knows his merge commits.  So there’s a reason for those.

LGTM

dscho commented:
> I wanted to test some redirects but I had the same local-only problem as last time I tried to serve the page locally: got infinite redirects when navigating.

@LemmingAvalanche how do you serve the local site? Via `hugo --serve`? That won't work, due to the `.html`-less pages that GitHub supports (and which we _must_ use for backwards compatibility with the Rails app version of git-scm.com).

Could you try serving it via `script/serve-public.js` instead?

dscho commented:
> I was a bit confused by all the one-commit merge commits for the localized book versions. But from reading the Git mailing list I know that Dscho knows his merge commits. So there’s a reason for those.

@LemmingAvalanche you mean commits like https://github.com/git/git-scm.com/pull/1915/commits/4b0016970bcd5ed5d01024d76d7343dbbe50cc8c? These are a side effect of [a matrix of jobs](https://github.com/git/git-scm.com/blob/c5766de01017595eeb1829a449ab9834a95ad777/.github/workflows/update-book.yml#L46-L49) taking care of updating the individual translations of the book. The changes made by those jobs all have to be merged together. (They could be merged via a single octopus merge, but my experience is that octopus merges baffle users and Git commands alike, therefore I avoid them.)

dscho commented:
Replying to [this thread](https://github.com/git/git-scm.com/pull/1915#discussion_r1836602885) more visibly:

> Uh oh. The `$v1_to_v2` data is not even used. Which means that https://dscho.github.io/git-scm.com/book/en/Git-Tools-Submodules redirects to https://dscho.github.io/git-scm.com/book/en/v2/GitHub-Summary (i.e. 6.6 -> 6.6, when it should redirect to https://dscho.github.io/git-scm.com/book/en/v2/Git-Tools-Submodules instead, i.e. to 7.11)!
> 
> Will fix immediately.

I fixed this, [force-pushed](https://github.com/git/git-scm.com/pull/1915#event-15250859805), then let the `update-book.yml` workflow do its job. The combined result is: https://github.com/git/git-scm.com/compare/4b0016970bcd5ed5d01024d76d7343dbbe50cc8c..6e803d156d69500d0cf099771078e46934d4cd3e

You can see that this fixed the English links starting here:

https://github.com/git/git-scm.com/compare/4b0016970bcd5ed5d01024d76d7343dbbe50cc8c..6e803d156d69500d0cf099771078e46934d4cd3e#diff-6bee9d14be43c9ee90083ef49c1b12679f64f40497a681f16996b354c6fb8703

You can verify that the Git-Tools-Submodules section I mentioned in the quoted message is now correctly aliased: https://github.com/git/git-scm.com/compare/4b0016970bcd5ed5d01024d76d7343dbbe50cc8c..6e803d156d69500d0cf099771078e46934d4cd3e#diff-7b2e103253875b7068636aa3fa1b96057a5fcd6e3f82ff78de2554dfc3652651

Or just direct your web browser to https://dscho.github.io/git-scm.com/book/en/Git-Tools-Submodules or to https://dscho.github.io/git-scm.com/book/en/v1/Git-Tools-Submodules to see that it works now.

dscho commented:
Will have to re-fix, hold on a sec.

dscho commented:
> Will have to re-fix, hold on a sec.

Fixed. Still had this left-over from an internal iteration: https://github.com/git/git-scm.com/compare/6e803d156d69500d0cf099771078e46934d4cd3e..8d9bfac33af659438239db9f9926f5da5181717d

LemmingAvalanche commented:
> (They could be merged via a single octopus merge, but my experience is that octopus merges baffle users and Git commands alike, therefore I avoid them.)

@dscho A sound assessment. :P

dscho commented:
> > I wanted to test some redirects but I had the same local-only problem as last time I tried to serve the page locally: got infinite redirects when navigating.
> 
> @LemmingAvalanche how do you serve the local site? Via `hugo --serve`? That won't work, due to the `.html`-less pages that GitHub supports (and which we _must_ use for backwards compatibility with the Rails app version of git-scm.com).
> 
> Could you try serving it via `script/serve-public.js` instead?

@LemmingAvalanche did you have any luck with that yet?

LemmingAvalanche commented:
> did you have any luck with that yet?

@dscho I’ll try this evening

LemmingAvalanche commented:
> I’ll try this evening

This setup works fine.

```
hugo
node script/serve-public.js
```

I’m writing a longer reply now.

LemmingAvalanche commented:
Most links are good to go.  I just checked that the title in the link makes sense with the title of the subchapter that I arrived at via the redirect.

See [Dscho’s findings in this comment](https://github.com/git/git-scm.com/issues/1782#issuecomment-2457407947). Maybe in particular the [Wayback Machine snapshot of](https://web.archive.org/web/20140102092829/https://git-scm.com/book/) the v1 book.

### Setup

```
hugo
node script/serve-public.js
```

### None of the chapter subtitle links work

Each v1, X.1 chapter has a link for the chapter title and one link for the chapter title plus the subtitle.  The chapter title (only) links but not the second variant.

For example:

- http://localhost:5000/book/en/Git-Branching
- Which correctly redirects to http://localhost:5000/book/en/v2/Git-Branching-Branches-in-a-Nutshell
- But: http://localhost:5000/book/en/Git-Branching-Branches-in-a-Nutshell
- Gets 404

The affected links are the following (space separated).  I just end up at the first link instead of the expected on.
```
Link Expected
localhost:5000/book/en/Getting-Started-About-Version-Control http://localhost:5000/book/en/v2/Getting-Started-About-Version-Control
localhost:5000/book/en/Git-Basics-Getting-a-Git-Repository http://localhost:5000/book/en/v2/Git-Basics-Getting-a-Git-Repository
localhost:5000/book/en/Git-Branching-What-a-Branch-Is http://localhost:5000/book/en/v2/Git-Branching-Branches-in-a-Nutshell
localhost:5000/book/en/Git-on-the-Server-The-Protocols http://localhost:5000/book/en/v2/Git-on-the-Server-The-Protocols
localhost:5000/book/en/Distributed-Git-Distributed-Workflows http://localhost:5000/book/en/v2/Distributed-Git-Distributed-Workflows
localhost:5000/book/en/Git-Tools-Revision-Selection http://localhost:5000/book/en/v2/Git-Tools-Revision-Selection
localhost:5000/book/en/Customizing-Git-Git-Configuration http://localhost:5000/book/en/v2/Customizing-Git-Git-Configuration
localhost:5000/book/en/Git-and-Other-Systems-Git-and-Subversion http://localhost:5000/book/en/v2/Git-and-Other-Systems-Git-as-a-Client
localhost:5000/book/en/Git-Internals-Plumbing-and-Porcelain http://localhost:5000/book/en/v2/Git-Internals-Plumbing-and-Porcelain
```

These are the links that have different titles in the two versions:

```
Link Expected
localhost:5000/book/en/Git-Branching-What-a-Branch-Is http://localhost:5000/book/en/v2/Git-Branching-Branches-in-a-Nutshell
localhost:5000/book/en/Git-and-Other-Systems-Git-and-Subversion http://localhost:5000/book/en/v2/Git-and-Other-Systems-Git-as-a-Client
```

### Chapter 2

Wrong redirect for this one:

```
Link Expected Actual
localhost:5000/book/en/Git-Basics-Tips-and-Tricks 404 http://localhost:5000/book/en/v2/Git-Basics-Git-Aliases
```

I don’t see any “tips and tricks”.  Seems like this should just 404.

### Chapter 4

I’ve added a column `404?` since the faulty behavior includes a 404. But three of them should result in 404, to be clear, just without the faulty redirect.

```
Link Expected Actual 404?
localhost:5000/book/en/Git-on-the-Server-GitWeb http://localhost:5000/book/en/v2/Git-on-the-Server-GitWeb http://localhost:1313/book/en/v2/Git-on-the-Server-Smart-HTTP Yes
localhost:5000/book/en/Git-on-the-Server-Gitosis 404 http://localhost:1313/book/en/v2/Git-on-the-Server-GitWeb Yes
localhost:5000/book/en/Git-on-the-Server-Gitolite 404 http://localhost:1313/book/en/v2/Git-on-the-Server-GitLab Yes
localhost:5000/book/en/Git-on-the-Server-Hosted-Git 404 http://localhost:1313/book/en/v2/Git-on-the-Server-Summary Yes
```

### Chapter 5

The summary page doesn’t work for me.

```
Link Expected Actual 404?
http://localhost:5000/book/en/Distributed-Git-Summary http://localhost:5000/book/en/v2/Distributed-Git-Summary http://localhost:1313/book/en/v2/Distributed-Git-Summary Yes
```

### Chapter 10

Here are wrong redirects for all subchapters.  The pattern is that you get redirected to the previous subchapter.  For example the first link redirects to 9.3 “Git and Other Systems - Summary”.  The last one (summary) redirects to the redirect page that we expected for the second-to-last v1 link.

```
Link Expected Actual
localhost:5000/book/en/Git-Internals-Git-References http://localhost:5000/book/en/v2/Git-Internals-Git-References http://localhost:1313/book/en/v2/Git-and-Other-Systems-Summary
localhost:5000/book/en/Git-Internals-Packfiles http://localhost:5000/book/en/v2/Git-Internals-Packfiles http://localhost:5000/book/en/v2/Git-Internals-Git-References
localhost:5000/book/en/Git-Internals-The-Refspec http://localhost:5000/book/en/v2/Git-Internals-The-Refspec http://localhost:5000/book/en/v2/Git-Internals-Packfiles
localhost:5000/book/en/Git-Internals-Transfer-Protocols http://localhost:5000/book/en/v2/Git-Internals-Transfer-Protocols http://localhost:5000/book/en/v2/Git-Internals-The-Refspec
localhost:5000/book/en/Git-Internals-Maintenance-and-Data-Recovery http://localhost:5000/book/en/v2/Git-Internals-Maintenance-and-Data-Recovery http://localhost:5000/book/en/v2/Git-Internals-Transfer-Protocols
localhost:5000/book/en/Git-Internals-Summary http://localhost:5000/book/en/v2/Git-Internals-Summary http://localhost:5000/book/en/v2/Git-Internals-Maintenance-and-Data-Recovery
```

dscho commented:
@LemmingAvalanche thank you for doing some well-needed QA!

> Each v1, X.1 chapter has a link for the chapter title and one link for the chapter title plus the subtitle. The chapter title (only) links but not the second variant.
> 
> For example:
> 
>     * http://localhost:5000/book/en/Git-Branching
> 
>     * Which correctly redirects to http://localhost:5000/book/en/v2/Git-Branching-Branches-in-a-Nutshell
> 
>     * But: http://localhost:5000/book/en/Git-Branching-Branches-in-a-Nutshell
> 
>     * Gets 404

I tried very hard to get both chapter X _and_ the X.1 section to link to X.1 (because there are no chapter links in v2). [Here](https://github.com/dscho/git-scm.com/blob/fb26f7f84fd48b22f406fda3a17aba25221d551b/data/book_v1.yml#L37) is where both v1's 3 and 3.1 are supposed to map to v2's 3.1.

But I see that https://dscho.github.io/git-scm.com/book/en/Git-Branching-What-a-Branch-Is does not redirect as desired...

&lt;clicketyclick&gt;

[Aha](https://stackoverflow.com/questions/5587264/do-end-vs-curly-braces-for-blocks-in-ruby#5587403)! My command of the Ruby language was insufficient, and I failed to realize that the way I implemented `book_v1_aliases`, an enumerator was returned instead of an array, which means that only the first array item (and in this case, intended alias) was respected.

This [is now fixed](https://github.com/git/git-scm.com/compare/8d9bfac33af659438239db9f9926f5da5181717d..9333518156997adf3842e7037aefc6628eefa127#diff-a426683fbe32cb8d1b5b7329662ccb9bde723e924dc8a64bdf1badef479f1876L215-R217).

> ### Chapter 2
> 
> Wrong redirect for this one:
> 
> ```
> Link Expected Actual
> localhost:5000/book/en/Git-Basics-Tips-and-Tricks 404 http://localhost:5000/book/en/v2/Git-Basics-Git-Aliases
> ```

Actually, if you have a closer look, https://web.archive.org/web/20141022015825/http://git-scm.com/book/en/Git-Basics-Tips-and-Tricks#Git-Aliases talks about Git Aliases, and the other half of the section talks about completion (which [I did not find mentioned in v2](https://github.com/search?q=repo%3Aprogit%2Fprogit2%20auto-completion&type=code)).

So I think that "Git Aliases" _is_ the best match here.

> ### Chapter 4
> 
> I’ve added a column `404?` since the faulty behavior includes a 404. But three of them should result in 404, to be clear, just without the faulty redirect.
> 
> ```
> Link Expected Actual 404?
> localhost:5000/book/en/Git-on-the-Server-GitWeb http://localhost:5000/book/en/v2/Git-on-the-Server-GitWeb http://localhost:1313/book/en/v2/Git-on-the-Server-Smart-HTTP Yes
> localhost:5000/book/en/Git-on-the-Server-Gitosis 404 http://localhost:1313/book/en/v2/Git-on-the-Server-GitWeb Yes
> localhost:5000/book/en/Git-on-the-Server-Gitolite 404 http://localhost:1313/book/en/v2/Git-on-the-Server-GitLab Yes
> localhost:5000/book/en/Git-on-the-Server-Hosted-Git 404 http://localhost:1313/book/en/v2/Git-on-the-Server-Summary Yes
> ```

I considered 404ing, but thought it would be much more helpful to redirect to http://localhost:5000/book/en/v2/Git-on-the-Server-Third-Party-Hosted-Options instead. 404s kind of suck when you're trying to find information.

> ### Chapter 5
> 
> The summary page doesn’t work for me.
> 
> ```
> Link Expected Actual 404?
> http://localhost:5000/book/en/Distributed-Git-Summary http://localhost:5000/book/en/v2/Distributed-Git-Summary http://localhost:1313/book/en/v2/Distributed-Git-Summary Yes
> ```

Thank you for pointing this out! This was an inadvertent omission of the section 5.4 in the mapping, which [is now fixed](https://github.com/git/git-scm.com/compare/8d9bfac33af659438239db9f9926f5da5181717d..9333518156997adf3842e7037aefc6628eefa127#diff-10ec3502d9f7e93f1127bfdfa052a90aaf8c56ab3811425c21498e4fd6b96cddL21-R21).

> ### Chapter 10
> 
> Here are wrong redirects for all subchapters. The pattern is that you get redirected to the previous subchapter. For example the first link redirects to 9.3 “Git and Other Systems - Summary”. The last one (summary) redirects to the redirect page that we expected for the second-to-last v1 link.
> 
> ```
> Link Expected Actual
> localhost:5000/book/en/Git-Internals-Git-References http://localhost:5000/book/en/v2/Git-Internals-Git-References http://localhost:1313/book/en/v2/Git-and-Other-Systems-Summary
> localhost:5000/book/en/Git-Internals-Packfiles http://localhost:5000/book/en/v2/Git-Internals-Packfiles http://localhost:5000/book/en/v2/Git-Internals-Git-References
> localhost:5000/book/en/Git-Internals-The-Refspec http://localhost:5000/book/en/v2/Git-Internals-The-Refspec http://localhost:5000/book/en/v2/Git-Internals-Packfiles
> localhost:5000/book/en/Git-Internals-Transfer-Protocols http://localhost:5000/book/en/v2/Git-Internals-Transfer-Protocols http://localhost:5000/book/en/v2/Git-Internals-The-Refspec
> localhost:5000/book/en/Git-Internals-Maintenance-and-Data-Recovery http://localhost:5000/book/en/v2/Git-Internals-Maintenance-and-Data-Recovery http://localhost:5000/book/en/v2/Git-Internals-Transfer-Protocols
> localhost:5000/book/en/Git-Internals-Summary http://localhost:5000/book/en/v2/Git-Internals-Summary http://localhost:5000/book/en
> ```

Thank you for diligent checking!

This was a mistake in my mapping, where I accidentally doubled the "10.2" entry, which [is now fixed](https://github.com/git/git-scm.com/compare/8d9bfac33af659438239db9f9926f5da5181717d..9333518156997adf3842e7037aefc6628eefa127#diff-10ec3502d9f7e93f1127bfdfa052a90aaf8c56ab3811425c21498e4fd6b96cddL25-R25).

LemmingAvalanche commented:
The following was tested on 933351815 (Merge branch 'book-fa' of bundle-fa/fa.bundle into book-v1-redirects, 2024-11-17)
.

> Actually, if you have a closer look, https://web.archive.org/web/20141022015825/http://git-scm.com/book/en/Git-Basics-Tips-and-Tricks#Git-Aliases talks about Git Aliases, and the other half of the section talks about completion (which [I did not find mentioned in v2](https://github.com/search?q=repo%3Aprogit%2Fprogit2%20auto-completion&type=code)).
>
> So I think that "Git Aliases" is the best match here.

@dscho 

Okay got it.  Makes sense.

> > Chapter 4
>
> ...
> I considered 404ing, but thought it would be much more helpful to redirect to http://localhost:5000/book/en/v2/Git-on-the-Server-Third-Party-Hosted-Options instead. 404s kind of suck when you're trying to find information.

Confirmed that this works as intended.  See below in the chapter walkthrough.

### Chapters 1–3

All okay.  Nothing to note.

### Chapter 4

Like Dscho mentioned here: some outdated-in-v2 links now point to

http://localhost:5000/book/en/v2/Git-on-the-Server-Third-Party-Hosted-Options

Which makes sense when we see that these are thirt-party things, just not
covered in v2:

localhost:5000/book/en/Git-on-the-Server-GitWeb
localhost:5000/book/en/Git-on-the-Server-Gitosis
localhost:5000/book/en/Git-on-the-Server-Gitolite

### Chapter 5

All okay.

### Chapter 6

v2 chapter 6 is on GitHub.  Does not exist in v1.

(It seems that I ended up using the v2 chapters as the headings since they don’t line up quite as I expected)

### Chapter 7

“Subtree merging” redirects to ch. 7.1:

```
Link Actual
localhost:5000/book/en/Git-Tools-Subtree-Merging http://localhost:5000/book/en/v2/Git-Tools-Revision-Selection
```

It might be better if it redirected to ch. 7.8:

http://localhost:5000/book/en/v2/Git-Tools-Advanced-Merging

Because that one has the subsubchapter (?) “Subtree Merging”.

### Chapter 8

All good.

### Chapter 9

v1 ch. 8.1 with title “Git and Other Systems: Git and Subversion” redirects to 9.1 “Git and Other Systems - Git as a Client”.

```
Link Actual
localhost:5000/book/en/Git-and-Other-Systems-Git-and-Subversion http://localhost:5000/book/en/v2/Git-and-Other-Systems-Git-as-a-Client
```

Which makes sense since:

1. Both are the first subchapter
2. There is a “Git and Subversion” subsubchapter/section in v2

### Chapter 10

All good.

> Thank you for diligent checking!

Thanks for the structured reply :D

dscho commented:
> ### Chapter 7
> “Subtree merging” redirects to ch. 7.1:
> 
> ```
> Link Actual
> localhost:5000/book/en/Git-Tools-Subtree-Merging http://localhost:5000/book/en/v2/Git-Tools-Revision-Selection
> ```
> 
> It might be better if it redirected to ch. 7.8:
> 
> http://localhost:5000/book/en/v2/Git-Tools-Advanced-Merging
> 
> Because that one has the subsubchapter (?) “Subtree Merging”.

[Fixed](https://github.com/git/git-scm.com/compare/9333518156997adf3842e7037aefc6628eefa127..47f14f07aa51a055bcbb3e006ff285858bd7ef5f)!

Of course it would be _even better_ if it redirected to https://git-scm.com/book/en/v2/Git-Tools-Advanced-Merging#_subtree_merge, but that would require more than just adding `aliases` to the target page; [Files with `redirect_to` front matter](https://github.com/dscho/git-scm.com/blob/5c9795e502abf709ee21a33d9643454f44667d03/script/book.rb#L162-L165) would need to be created, similar to the redirects for `/book/en` and `/book/en/v1` to `/book/en/v2`. Doable, but I'd like to declare the current state as good enough. Agree?

LemmingAvalanche commented:
> Fixed!

Confirmed.

> Doable, but I'd like to declare the current state as good enough. Agree?

Yep.

LGTM
